File: internal/redis/fixed_window.go
package redis

import (
	"context"
	_ "embed"
	"fmt"
	"strconv"
)

//go:embed scripts/fixed_window.lua
var fixedWindowScript string

type FixedWindowResult struct {
	Allowed      bool
	Remaining    int
	Limit        int
	RetryAfterMs int64
}

func CheckFixedWindow(
	ctx context.Context,
	client *Client,
	key string,
	limit int,
	windowSeconds int,
) (FixedWindowResult, error) {

	if limit <= 0 {
		return FixedWindowResult{}, fmt.Errorf(
			"fixed window limit must be greater than zero",
		)
	}

	if windowSeconds <= 0 {
		return FixedWindowResult{}, fmt.Errorf(
			"fixed window window_seconds must be greater than zero",
		)
	}

	result, err := client.client.Eval(
		ctx,
		fixedWindowScript,
		[]string{key},
		strconv.Itoa(limit),
		strconv.Itoa(windowSeconds),
	).Result()

	if err != nil {
		return FixedWindowResult{}, fmt.Errorf(
			"execute fixed window script: %w",
			err,
		)
	}

	values, ok := result.([]interface{})
	if !ok || len(values) != 4 {
		return FixedWindowResult{}, fmt.Errorf(
			"unexpected fixed window result",
		)
	}

	allowed, ok := values[0].(int64)
	if !ok {
		return FixedWindowResult{}, fmt.Errorf(
			"invalid fixed window allowed result",
		)
	}

	remaining, ok := values[1].(int64)
	if !ok {
		return FixedWindowResult{}, fmt.Errorf(
			"invalid fixed window remaining result",
		)
	}

	resultLimit, ok := values[2].(int64)
	if !ok {
		return FixedWindowResult{}, fmt.Errorf(
			"invalid fixed window limit result",
		)
	}

	retryAfterMs, ok := values[3].(int64)
	if !ok {
		return FixedWindowResult{}, fmt.Errorf(
			"invalid fixed window retry_after result",
		)
	}

	return FixedWindowResult{
		Allowed:      allowed == 1,
		Remaining:    int(remaining),
		Limit:        int(resultLimit),
		RetryAfterMs: retryAfterMs,
	}, nil
}

File: internal/redis/scripts/fixed_window.lua
local key = KEYS[1]

local limit = tonumber(ARGV[1])
local window_seconds = tonumber(ARGV[2])

if not limit or not window_seconds then
    return redis.error_reply("invalid fixed window arguments")
end

if limit <= 0 or window_seconds <= 0 then
    return redis.error_reply(
        "fixed window arguments must be greater than zero"
    )
end

local current_count = tonumber(redis.call("GET", key))

if not current_count then
    current_count = 0
end

local allowed = 0
local remaining = 0

if current_count < limit then
    current_count = current_count + 1
    allowed = 1
    remaining = limit - current_count
else
    remaining = 0
end

if allowed == 1 then
    redis.call(
        "SET",
        key,
        current_count,
        "EX",
        window_seconds
    )
end

local ttl_ms = redis.call("PTTL", key)

local retry_after_ms = 0

if allowed == 0 then
    retry_after_ms = ttl_ms
end

return {
    allowed,
    remaining,
    limit,
    retry_after_ms
}